locations/storefinders/nomnom.py

```diff
@@ -0,0 +1,92 @@
+import re
+from datetime import date, timedelta
+from typing import Iterable
+from urllib import parse as urlparse
+
+from scrapy import Spider
+from scrapy.http import Request, Response, TextResponse
+
+from locations.categories import Extras, apply_yes_no
+from locations.dict_parser import DictParser
+from locations.hours import DAYS, DAYS_3_LETTERS, OpeningHours
+from locations.items import Feature
+
+
+def slugify(s: str) -> str:
+    return re.sub(r"\s+", "-", s).casefold()
+
+
+class NomNomSpider(Spider):
+    """NomNom is an "accelerator technology" by Bounteous for restaurants and
+    retailers:
+    https://www.bounteous.com/industries/restaurant-convenience/nomnom/
+    To use, specify "domain" as the base domain of the website."""
+
+    domain: str | None = None
+    use_calendar: bool = True
+
+    def _append_calendar_param(self, url):
+        if self.use_calendar:
+            parsed = urlparse.urlparse(url)
+            params = urlparse.parse_qs(parsed.query)
+            today = date.today()
+            params["nomnom"] = "calendars"
+            params["nomnom_calendars_from"] = [today.strftime("%Y%m%d")]
+            params["nomnom_calendars_to"] = [(today + timedelta(days=6)).strftime("%Y%m%d")]
+            parsed = parsed._replace(query=urlparse.urlencode(params, doseq=True))
+            return urlparse.urlunparse(parsed)
+        else:
+            return url
+
+    def start_requests(self) -> Iterable[Request]:
+        if self.start_urls:
+            for url in self.start_urls:
+                yield Request(self._append_calendar_param(url), dont_filter=True)
+        elif self.domain:
+            yield Request(self._append_calendar_param(f"https://nomnom-prod-api.{self.domain}/restaurants/"))
+
+    @staticmethod
+    def parse_opening_hours(calendar: dict) -> OpeningHours:
+        oh = OpeningHours()
+        for row in calendar["ranges"]:
+            if row["weekday"] not in DAYS_3_LETTERS:
+                continue
+            day = DAYS[DAYS_3_LETTERS.index(row["weekday"])]
+            oh.add_range(day, row["start"], row["end"], time_format="%Y%m%d %H:%M")
+        return oh
+
+    CALENDAR_KEYS = {
+        "drivethru": "opening_hours:drive_through",
+        "dispatch": "delivery",
+    }
+
+    def parse_locations(self, response: TextResponse) -> Iterable[Feature]:
+        for location in response.json()["restaurants"]:
+            item = DictParser.parse(location)
+            item["ref"] = location["extref"]
+            item["branch"] = location["name"].replace(location["storename"], "").strip()
+            item["name"] = location["storename"]
+
+            apply_yes_no(Extras.DELIVERY, item, location["candeliver"] or location["supportsdispatch"])
+            apply_yes_no(Extras.TAKEAWAY, item, location["canpickup"] or location["supportscurbside"])
+            apply_yes_no(Extras.DRIVE_THROUGH, item, location["supportsdrivethru"])
```

Comment:
All these brands appear to block out-of-country access so I can't check myself--should these have `, False` added to the end of the parameters list if the source data is always `True` or `False`? The code you have hear indicates it's unknown whether a location *does not* have a drive through--we'd only ever know if it *does* have a drive through.

Reply:
It seems to me like this software solution was built primarily as an online-ordering platform, with the store finder only built because you need to be able to select a store to order from. So I think these attributes are used to tell whether that mode is available for online ordering. So, if `supportsdrivethru` is `True`, for example, that means that you can order online to pick up at the drive-through. They might still have a drive through that doesn't allow online ordering.